exporter/prometheusremotewriteexporter/README.md

```diff
@@ -36,6 +36,9 @@ The following settings can be optionally configured:
 - `headers`: additional headers attached to each HTTP request. 
   - *Note the following headers cannot be changed: `Content-Encoding`, `Content-Type`, `X-Prometheus-Remote-Write-Version`, and `User-Agent`.*
 - `namespace`: prefix attached to each exported metric name.
+- `queue_config`: fine tuning for the outgoing request queues.
+  - `size`: number of OTLP metrics that can be queued.
+  - `min_shards`: minimum number of shards to use to fan out the outgoing requests.
```

Comment:
The min_shards is not implemented correct? It does not do sharding or am I missing something?

Reply:
Min shards controls remote_write fanout: https://github.com/open-telemetry/opentelemetry-collector/pull/3046/files#diff-2631d8a8855fdc08cf077c0ab99544147ca73eec572bc871d8fe717e826eda38R279-R298

From Prometheus:

```
      |-->  queue (shard_1)   --> remote endpoint
WAL --|-->  queue (shard_...) --> remote endpoint
      |-->  queue (shard_n)   --> remote endpoint
```

"Min shards configures the minimum number of shards used by Prometheus, and is the number of shards used when remote write starts. If remote write falls behind, Prometheus will automatically scale up the number of shards so most users do not have to adjust this parameter. However, increasing min shards will allow Prometheus to avoid falling behind at the beginning while calculating the required number of shards."

This is very similar to the fanout we have in the exporter. We haven't implemented max_shards yet to automatically increase the number of shards, but this PR is a good first step to give users a way to fine tune.

Reply:
I do understand what PRW min_shards does, but in this PR min_shards configures the concurrency, but does not split the incoming requests (a.k.a "sharding") to ensure that for one worker we always get the same shard (a.k.a I think we just add back the out of order problem?)

Probably I am missing something, but from the code in this PR this is what I am seeing, probably there are other internal things that happen, if that is the case please point me to that code.

Reply:
batchTimeSeries shard the samples by timeseries. https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/prometheusremotewriteexporter/exporter.go#L273 Each request only contains samples from a particular timeseries already. This PR is only making something hardcoded a configurable setting.

Reply:
Maybe, I'm misreading your comment here. If you are asking if we are implementing the shards and queues exactly the way Prometheus does, we don't. But this is something we can start with until we have time to replicate their queuing mechanism or reuse their code.

Reply:
>  If you are asking if we are implementing the shards and queues exactly the way Prometheus does, we don't. 

@rakyll this is what I was asking

If we don't do, I know consistency in terms of names is good, but maybe add a caveat that the full functionality is not implemented, and for the moment this only controls the number of workers.

Reply:
> and for the moment this only controls the number of workers

This is what min_shards in Prometheus does.

min_shards controls the number of shards being concurrently sent to a remote write endpoint.  We create n requests sharded by timeseries name. Then, we create up to min_shards workers to send remote writes. This is not different from what Prom is doing with min_shards. 

We just don't autoscale the workers up to max_shards. This is ok, we will error if users want to bring their existing config with max_shard. 

If we want to call this setting something else like "workers", we will have to deprecate it once we have min_shards and max_shards implemented.

Reply:
The difference I forgot to mention in the previous comment is Prom always has min_shards workers available. We create up to min_shards shards when we have metrics to export. The workers are not continuously running. This is only a deal when there is too much congestion on the "export" method. We should fix the mechanism here either now and as a follow up PR.

Reply:
Renamed min_shards to concurrency. PTAL.